### PR TITLE
Decorator to cache fixtures, meant for manual debugging mostly

### LINKED ISSUE #7227: caching fixtures for easier debugging
I'm creating a test which depends on a setup (via `@pytest.fixture(scope='session')`) which takes, say, 15 minutes to run. That is quite expensive to learn that I have made another error or I have not used some object correctly. And on each try, one more organization and location and domain and ... is created in my satellite.

Yes, I could use debugger and breakpoint or some interactive console and maybe everybody is happy with that when creating new test (?). I have created this decorator:

```
+import os
+from functools import wraps
+import pickle
+
+def debug_caching(func):
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        import os
+        # If debug caching is not enabled, simply just run the wrapped function and return
+        if 'DEBUG_CACHING' not in os.environ:
+            LOGGER.debug('Debug caching not enabled, just running %s' % func.__name__)
+            return func(*args, **kwargs)
+
+        # If debug caching is enabled
+        try:
+            with open('/tmp/debug_caching_data.pickle', 'rb') as fp:
+                debug_caching_data = pickle.load(fp)
+        except FileNotFoundError:
+            debug_caching_data = {}
+        LOGGER.debug('Loaded debug cache with %d items' % len(debug_caching_data))
+
+        if func.__name__ in debug_caching_data:
+            LOGGER.debug('Returning object %s from debug cache' % func.__name__)
+            return debug_caching_data[func.__name__]
+        else:
+            LOGGER.debug('Running %s and saving to debug cache' % func.__name__)
+            result = func(*args, **kwargs)
+            debug_caching_data[func.__name__] = result
+            with open('/tmp/debug_caching_data.pickle', 'wb') as fp:
+                pickle.dump(debug_caching_data, fp)
+            return result
+
+    return wrapper
```

and added this to all the fixtures I'm using:

```
 @pytest.fixture(scope='session')
+@debug_caching
 def user_credentials():
     # Create a new user with admin permissions
     login = gen_string('alphanumeric')
@@ -64,6 +102,7 @@ def user_credentials():
 
 
 @pytest.fixture(scope='session')
+@debug_caching
 def org(user_credentials):
     org = entities.Organization(user_credentials).create()
     with manifests.clone() as manifest:
@@ -72,6 +111,7 @@ def org(user_credentials):
 
 
 @pytest.fixture(scope='session')
+@debug_caching
 def loc(user_credentials, org):
     loc = entities.Location(user_credentials).create()
     loc.organization = [org]
```

It runs the fixture and saves its return value to a pickle file on a first run. Then, on every subsequent run it only reads that object from the pickle and runs mine test with that, so I can run the test in about second.

Does anybody else thinks it could be useful? Shall I/we invest more in exploring this path, or is this a bad practice with simple and better workaround?

### PR DESCRIPTION
**Pull request by jhutar**
Fixes #7227

Meant for manual debugging and developing tests with many or expensive fixtures which are not subject of development. Please see decorator's doc-string.

This commit just adds the decorator and it is up to user to enable it when needed. This commit does not change Robottelo or tests in any way.

### CODE REVIEW

File: robottelo/decorators/__init__.py [hunk 1]
```diff
@@ -376,3 +377,63 @@ def decorator(func):
...
+
+    :param func: function
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        import os
+        # If debug caching is not enabled, simply just run the wrapped function and return
+        if 'DEBUG_CACHING' not in os.environ:
+            LOGGER.debug('Debug caching not enabled, just running %s' % func.__name__)
```
Comment (JacobCallahan) [inline robottelo/decorators/__init__.py]: Ahh caught this after the issue. As I commented on that, you'll need to change your string formatting to use .format.

**Comment by vijay8451** [comment]
@jhutar ping

**Comment by vijay8451** [comment]
@jhutar ping , incase you missed last comment.

**Comment by JacobCallahan** [comment]
@SatelliteQE/quality-engineers does anyone want to take this up? If not, this is a candidate for closure.

**Comment by jhutar** [comment]
Sorry Vijay and Jake, I have missed the conversation. I do not plan to work on this now. Thank you!

**Comment by mirekdlugosz** [comment]
Looks like the only thing blocking that is that it uses `%` instead of `.format()` (or f-string) for formatting? I personally wouldn't reject PR on something as minute.

Having said that, this tries to solve real problem and help speed up test development. Test setup is potentially expensive (registering hosts, syncing repos, publishing content views) and it makes a lot of sense to allow test developer to spend that cost only once. I don't care about this particular implementation, but the general idea should definitely be there on automation group roadmap.

**Comment by JacobCallahan** [comment]
@mirekdlugosz I do care, as consistency is important for projects with more than a few contributors.
I'd be fine with this waiting until [1] is done, then @jhutar can use f-strings in the updated version.
[1] - https://github.com/SatelliteQE/robottelo-ci/pull/1768

**Comment by JacobCallahan** [comment]
Triaged by the review team. @rplevka has volunteered to take this on when he is able.

**Comment by omkarkhatavkar** [comment]
Can one of the admins verify this patch?

**Comment by latran** [comment]
@rplevka any updates on this?

**Comment by JacobCallahan** [comment]
Closing due to inactivity. This can be reopened if picked back up in the future.

**Pull request closed without merging**